www/docs/client/react/setup.mdx

```diff
@@ -97,6 +97,27 @@ export function App() {
 The reason for using `useState` in the creation of the `queryClient` and the `TRPCClient`, as opposed to declaring them outside of the component, is to ensure that each request gets a unique client when using SSR. If you use client side rendering then you can move them if you wish.
 :::
 
+:::tip
+React Query uses a technique called [structural sharing](https://tanstack.com/query/latest/docs/framework/react/guides/render-optimizations#structural-sharing) to ensure that as many references as possible will be kept intact between re-renders.
```

Comment:
I still want an explanation of why it's important to keep values intact between re-renders

I've personally haven't ever felt the need of this optimization and I refetch aggressively [on any mutation](https://trpc.io/docs/client/react/useUtils#invalidate-full-cache-on-every-mutation)

Reply:
I think there's two ways to look at it:

Say you have a page with an "edit" form that uses RHF `useForm` and fetches the current data client-side. There's multiple ways to do this, but there's a chance you have something like

```tsx
useEffect(() => {reset(data.values)}, [data.values])
```

somewhere.

With structural sharing, even if the query is refetched for some reason (for example, you switch tabs with the default behavior), the identity of `data.values` is stable as long at the data stays the same. Without structural sharing, the identity is not stable, the form is reset and you lose any changes you were in the process of making.
I think there are other ways to tackle this specific use-case, but overall I think it's a nice thing that helps react-query work in a more intuitive way.

The second argument is that structural sharing is already present in RQ by default, for json-compatible structures. Since trpc has a RQ integration, I think it makes sense to also provide this feature for those that want it, keeping behavior similar to vanilla RQ while benefiting from the type safety of trpc.
FWIW, stability for query results that contain `Date` objects has been requested / tried [multiple](https://github.com/TanStack/query/pull/6481) [times](https://github.com/TanStack/query/pull/7151) at RQ.